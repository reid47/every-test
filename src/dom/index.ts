import { RenderContainer } from '../RenderContainer';

const activeContainers: RenderContainer[] = [];

/**
 * Renders a DOM element and returns a render container wrapping
 * that element.
 *
 * @param element DOM node or HTML string representing DOM node
 * @param options Render options
 */
export const render = (element: HTMLElement | string, options = {}): RenderContainer => {
  const container = new RenderContainer(options);

  if (typeof element === 'string') {
    container.domNode.innerHTML = element;
  } else {
    container.domNode.appendChild(element);
  }

  activeContainers.push(container);
  container.mount();

  return container;
};

/**
 * Unmounts all remaining render containers from the DOM.
 */
export const cleanup = () => {
  while (activeContainers.length) {
    const container = activeContainers.pop();
    container && container.unmount();
  }
};
