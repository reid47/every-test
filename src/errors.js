export const incorrectNodeCount = (helperName, selector, amount) => {
  return new Error(
    `${helperName}: expected to find exactly one node matching '${selector}', but found ${amount}.`
  );
};

export const documentNotAvailable = () => {
  return new Error(
    `Either 'document' is undefined, or 'document.createElement' is not a function. A DOM must be available to render.`
  );
};

export const noRendersFound = componentType => {
  return new Error(
    `getPropsOf: No rendered instances of '${componentType}' found in component tree.`
  );
};

export const multipleRendersFound = componentType => {
  return new Error(
    `getPropsOf: Multiple rendered instances of '${componentType}' found in component tree. Maybe you want to use 'allPropsOf'?`
  );
};
