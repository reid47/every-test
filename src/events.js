export default function createEventHelpers({ findOnly }) {
  return {
    blurOn: selector => {
      findOnly(selector, 'blurOn').blur();
    },

    changeValueOf: (selector, newValue, options = {}) => {
      const node = findOnly(selector, 'changeValue');
      node.value = newValue;
      node.dispatchEvent(
        new Event('input', {
          bubbles: true,
          cancelable: true,
          ...options
        })
      );
    },

    clickOn: selector => {
      findOnly(selector, 'clickOn').click();
    },

    doubleClickOn: (selector, options = {}) => {
      findOnly(selector, 'doubleClickOn').dispatchEvent(
        new MouseEvent({
          bubbles: true,
          cancelable: true,
          ...options
        })
      );
    },

    focusOn: selector => {
      findOnly(selector, 'focusOn').focus();
    }
  };
}
