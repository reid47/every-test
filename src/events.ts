export const dispatch = (element, event) => {
  element.dispatchEvent(event);
};

export const dispatchChange = (element, newValue, options = {}) => {
  const { set: valueSetter }: PropertyDescriptor =
    Object.getOwnPropertyDescriptor(element, 'value') || {};

  const { set: prototypeValueSetter }: PropertyDescriptor =
    Object.getOwnPropertyDescriptor(Object.getPrototypeOf(element), 'value') || {};

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, newValue);
  } else if (valueSetter) {
    valueSetter.call(element, newValue);
  }

  const event = new Event('input', { bubbles: true, ...options });
  element.dispatchEvent(event);
};
