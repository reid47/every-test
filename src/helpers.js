import prettyFormat from 'pretty-format';

const { DOMElement, DOMCollection } = prettyFormat.plugins;

export const formatDomNode = (node, printOptions) => {
  return prettyFormat(node, {
    plugins: [DOMElement, DOMCollection],
    maxDepth: 24,
    ...printOptions
  });
};
