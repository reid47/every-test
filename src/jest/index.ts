import * as baseMatchers from '../matchers';
import { formatValue } from '../matchers/common';

function adaptMatcherResult({ matcherName, pass, message, received }) {
  const fullMatcherName = `${pass ? '.not' : ''}.${matcherName}`;

  const messageString = [
    `${this.utils.matcherHint(fullMatcherName, formatValue(received))}`,
    '',
    message
  ].join('\n');

  return { pass, message: () => messageString };
}

export const matchers = Object.keys(baseMatchers).reduce((acc, matcherName) => {
  const matcherFunction = baseMatchers[matcherName];

  return {
    ...acc,
    [matcherName](received, ...args) {
      const { pass, message } = matcherFunction(received, ...args);
      return adaptMatcherResult.call(this, { matcherName, received, pass, message });
    }
  };
}, {});
