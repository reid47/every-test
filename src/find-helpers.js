import { incorrectNodeCount } from './errors';

export default function createFindHelpers(domNode) {
  const find = selector => domNode.querySelector(selector);

  const findAll = selector => Array.from(domNode.querySelectorAll(selector));

  const findOnly = (selector, helperName = 'findOnly') => {
    const nodes = domNode.querySelectorAll(selector);
    if (nodes.length !== 1) throw incorrectNodeCount(helperName, selector, nodes.length);
    return nodes[0];
  };

  const findAllByText = toMatch => {
    return [...domNode.querySelectorAll('*')].filter(node => {
      const text = node.textContent;
      if (typeof toMatch === 'string') return text === toMatch;
      if (toMatch instanceof RegExp) return toMatch.test(text);
    });
  };

  const findByText = toMatch => findAllByText(toMatch)[0] || null;

  const findOnlyByText = toMatch => {
    const nodes = findAllByText(toMatch);
    if (nodes.length !== 1) throw incorrectNodeCount('findOnlyByText', toMatch, nodes.length);
    return nodes[0];
  };

  return { find, findAll, findOnly, findByText, findAllByText, findOnlyByText };
}
