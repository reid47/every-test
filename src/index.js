import React from 'react';
import ReactDOM from 'react-dom';
import createFindHelpers from './find-helpers';
import createEventHelpers from './event-helpers';
import createReactHelpers from './react-helpers';
import { documentNotAvailable } from './errors';

const containers = [];

export const cleanup = () => {
  while (containers.length) {
    const node = containers[containers.length - 1];
    ReactDOM.unmountComponentAtNode(node);
    document.body.removeChild(node);
    containers.pop();
  }
};

const cloneElementWithStubs = (element, shallowTypes) => {
  if (shallowTypes.has(element.type)) return null;

  return React.cloneElement(element, {
    children: React.Children.map(element.props.children, child => {
      return cloneElementWithStubs(child, shallowTypes);
    })
  });
};

export const render = (element, options = {}) => {
  if (typeof document !== 'object' || typeof document.createElement !== 'function') {
    throw documentNotAvailable();
  }

  const domNode = document.createElement('div');
  document.body.appendChild(domNode);

  const renderedElement =
    options.shallow && options.shallow.length
      ? cloneElementWithStubs(element, new Set(options.shallow))
      : element;

  ReactDOM.render(renderedElement, domNode);
  containers.push(domNode);

  const findHelpers = createFindHelpers(domNode);
  const eventHelpers = createEventHelpers(findHelpers);
  const reactHelpers = createReactHelpers(element);

  return {
    ...findHelpers,
    ...eventHelpers,
    ...reactHelpers,
    unmount: () => ReactDOM.unmountComponentAtNode(domNode),
    print: () => console.log(domNode.outerHTML)
  };
};
