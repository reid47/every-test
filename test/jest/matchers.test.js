import { matchers } from '../../src/jest';

describe('Jest matchers', () => {
  beforeAll(() => {
    expect.extend(matchers);
  });

  test('toHaveAttr', () => {
    const div = document.createElement('div');
    div.setAttribute('id', 'hello');

    expect(div).toHaveAttr('id', 'hello');
    expect(div).not.toHaveAttr('id', 'wow');
    expect(() => expect(div).toHaveAttr('id', 'wow')).toThrowErrorMatchingSnapshot();
    expect(() => expect(div).not.toHaveAttr('id', 'hello')).toThrowErrorMatchingSnapshot();
  });

  test('toHaveValue', () => {
    const input = document.createElement('input');
    input.value = 47;

    expect(input).toHaveValue('47');
    expect(input).not.toHaveValue('48');
    expect(input).not.toHaveValue(47);
    expect(() => expect(input).toHaveValue(47)).toThrowErrorMatchingSnapshot();
    expect(() => expect(input).not.toHaveValue('47')).toThrowErrorMatchingSnapshot();
  });
});
