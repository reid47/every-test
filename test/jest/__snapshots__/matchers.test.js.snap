// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest matchers toHaveAttr 1`] = `
"expect(<div id=\\"hello\\"></div>).toHaveAttr(...)

Expected node to have attribute \\"id\\" with value \\"wow\\", but actual value was \\"hello\\"."
`;

exports[`Jest matchers toHaveAttr 2`] = `
"expect(<div id=\\"hello\\"></div>).not.toHaveAttr(...)

Expected node not to have attribute \\"id\\" with value \\"hello\\"."
`;

exports[`Jest matchers toHaveValue 1`] = `
"expect(<input>).toHaveValue(...)

Expected node to have value 47, but actual value was \\"47\\"."
`;

exports[`Jest matchers toHaveValue 2`] = `
"expect(<input>).not.toHaveValue(...)

Expected node not to have value \\"47\\"."
`;
